safekeeper/src/wal_backup.rs

```diff
@@ -0,0 +1,473 @@
+        let rs = conf.backup_storage.as_ref().map(|c| {
+            GenericRemoteStorage::new(conf.timeline_dir(timeline_id), c)
+                .expect("Failed to create remote storage")
+        });
+
+        Box::new(rs)
+    });
+
+    let election_name = broker::get_campaign_name(
+        BACKUP_ELECTION_NAME.to_string(),
+        conf.broker_etcd_prefix.clone(),
+        timeline_id,
+    );
+    let my_candidate_name = broker::get_candiate_name(conf.my_id);
+    let election = broker::Election::new(
+        election_name,
+        my_candidate_name,
+        conf.broker_endpoints.clone(),
+    );
+
+    runtime.spawn(
+        backup_task(
+            backup_start,
+            *timeline_id,
+            conf.timeline_dir(timeline_id),
+            wal_seg_size_receiver,
+            lsn_committed_receiver,
+            shutdown_receiver,
+            lsn_backed_up_sender,
+            election,
+        )
+        .instrument(info_span!("Wal Backup", timeline_id.timeline_id = %timeline_id)),
+    );
+
+    Ok(WalBackup::new(
+        wal_seg_size_sender,
+        lsn_committed_sender,
+        lsn_backed_up_receiver,
+        shutdown_sender,
+    ))
+}
+
+pub fn create_noop() -> WalBackup {
+    let (lsn_committed_sender, _lsn_committed_receiver) = watch::channel(Lsn::INVALID);
+    let (_lsn_backed_up_sender, lsn_backed_up_receiver) = watch::channel(Lsn::INVALID);
+    let (wal_seg_size_sender, _wal_seg_size_receiver) = watch::channel::<u32>(0);
+    let (shutdown_sender, _shutdown_receiver) = watch::channel::<bool>(false);
+
+    WalBackup::new(
+        wal_seg_size_sender,
+        lsn_committed_sender,
+        lsn_backed_up_receiver,
+        shutdown_sender,
+    )
+}
+
+#[derive(Debug, Copy, Clone)]
+pub struct Segment {
+    seg_no: XLogSegNo,
+    start_lsn: Lsn,
+    end_lsn: Lsn,
+}
+
+impl Segment {
+    pub fn new(seg_no: u64, start_lsn: Lsn, end_lsn: Lsn) -> Self {
+        Self {
+            seg_no,
+            start_lsn,
+            end_lsn,
+        }
+    }
+
+    pub fn object_name(self) -> String {
+        XLogFileName(PG_TLI, self.seg_no, self.size())
+    }
+
+    pub fn file_path(self, timeline_dir: &Path) -> Result<PathBuf> {
+        Ok(timeline_dir.join(self.object_name()))
+    }
+
+    pub fn size(self) -> usize {
+        (u64::from(self.end_lsn) - u64::from(self.start_lsn)) as usize
+    }
+}
+
+fn get_segments(start: Lsn, end: Lsn, seg_size: usize) -> Vec<Segment> {
+    let first_seg = start.segment_number(seg_size);
+    let last_seg = end.segment_number(seg_size);
+
+    let res: Vec<Segment> = (first_seg..last_seg)
+        .map(|s| {
+            let start_lsn = XLogSegNoOffsetToRecPtr(s, 0, seg_size);
+            let end_lsn = XLogSegNoOffsetToRecPtr(s + 1, 0, seg_size);
+            Segment::new(s, Lsn::from(start_lsn), Lsn::from(end_lsn))
+        })
+        .collect();
+
+    res
+}
+
+static REMOTE_STORAGE: OnceCell<Box<Option<GenericRemoteStorage>>> = OnceCell::new();
```

Comment:
I doubt this should be *sync* OnceCell, and don't see why it is enum instead of trait -- that forces the manual `match` below.